### ISSUE TITLE
What should 'selected_values' do?

### ISSUE THREAD

**Issue by mikeorr** [mikeorr/WebHelpers2#19]
Jeff Dairiki:

I think the handling of `selected_values` is fairly broken. It only affects options which make it this far. Options within optgroups, or those which were already Option or OptGroup instances will not have there .selected attributes updated.

I'm not sure what the best fix is. One idea is to add a method, say set_selected, to Options, OptGroup, and maybe event Option. This method would take a single argument, selected_values, and would return a deep copy of self with the selected attributes of the descendant Options appropriately updated. (The selected_values argument of the Options constructor could probably be deleted, in this case.) If you like this idea, I can work up a pull request. Let me know.

Regarding `Options.__init__()`, version 457ef29, lines 341-369.

**Comment by mikeorr** [comment]
See what you think of this: f714de4 (master, 2015-01-03).

I made the `options` argument more restrictive, similar to your suggestion, and stopped doing the complex parsing of tuples and embedded lists.

The preselection API has also changed. `options.render(selected_values=None)` replaces the earlier stringification. `option.selected` is gone, so there's no need for a `with_selected` method.

I want to get the final done soon, so let me know if you think this is good enough for an RC3.

**Comment by dairiki** [comment]
Looks good to me!

Just one nit: for convenience, the `OptGroup` constructor should take an optional `options` argument, so that one can construct an `OptGroup` in place:
```python
mygroup = OptGroup('Fruits', ['Apple', 'Pear', Option('Tomato', 'veggiefruit')])
```
Unless I hear otherwise, I will prepare a pull request for this tomorrow.

**Comment by mikeorr** [comment]
I was debating whether that would be worthwhile. I'm not especially fond of it because it would have to: 
* parse the elements itself (bad and duplicative)
* call back to Options and take the resulting list (a misuse of Options, and it might call back to OptGroup. The triangle of recursive calls between OptGroup, Options, and Option might break in edge cases. That was one of the reasons I ditched the RC2 implementation.)
* extract the parsing code to a base method.

The latter sounds best. Both are subclasses of _OptionsList, and in both cases they're taking an iterable argument and appending to self. And if the base method calls both Option and OptGroup, that's not as error-prone as Options and OptGroup calling each other.

**Comment by dairiki** [comment]
Re: The latter sounds best

Agreed.  Option 3 was, more or less, my plan.  I was planning on adding a static method `_OptionsList._coerce_options(options)` that the contructors for `OptGroup` and `Options` would use to massage the constructor argument to a usable list of options.  (I *think* that `_coerce_options` doesn't have to make any recursive calls.  The only thing it needs to construct is `Option`s.)

All `_OptionsList`s are sequences of options.  It makes sense to me that they should all be directly constructable as such.  (The current implementation of `OptGroup.__repr__` seems to agree, for what it's worth.)  Also, in cases where an `OptGroup` is a compile-time constant, it would be nice to be able to declare it as such, rather than be forced to construct it imperatively.

**Comment by dairiki** [comment]
See PR #21

**Comment by mikeorr** [comment]
I tried your code but decided to stick with my idea of an ``OptionsList._parse_options`` method that parses the entire iterator and adds the options. Released 2.0rc3.
